spl/labels.py

```diff
@@ -186,55 +198,90 @@ def __get_application_numbers(self, set_id, bs_content):
+            # Process the label text; normalize remove nbsp
+            text = unicodedata.normalize("NFKC", text.lstrip().rstrip())
+            text = re.sub(r"\n\s*\n", "\n", text)
+            return text
+
+        def get_xml_title(text):
+            # strip extra \n, \t, and spaces in titles, but keep number
+            text = get_xml_text(text)
+            text = re.sub(r"\.\s+", ".", text)
+            text = re.sub(r"\s+", " ", text)
+            return text
 
         # Get and process all "title" tags
         titles = bs_content.find_all("title")
         labels = []
 
-        for title in titles:
-            if isinstance(title, Tag):
-                for section_title in SplHistoricalLabels.LABEL_SECTIONS:
-                    if section_title in title.text:
-                        # Match the title tags in the sub-section
-                        sub_titles = title.parent.find_all("title")
-                        sub_sections = []
-                        for sub_title in sub_titles:
-                            if isinstance(sub_title, Tag):
-                                for (
-                                    sub_section_title
-                                ) in SplHistoricalLabels.LABEL_SECTIONS[
-                                    section_title
-                                ]:
-                                    if sub_section_title in sub_title.text:
-                                        # Extract the sub-title text
-                                        sub_sections.append(
-                                            {
-                                                "name": sub_section_title,
-                                                "text": get_xml_text(
-                                                    sub_title.parent
-                                                ),
-                                            }
-                                        )
-                                        # Remove the sub title from the main title
-                                        sub_title.parent.decompose()
-
-                        # Add the section text, along with the sub-sections
+        i = 0
+        while i < len(titles):
+            title = titles[i]
+            # test if any substring of title is in LABEL_SECTION
+            title_substring_in_LABEL_SECTION = any(
+                label_section.lower() in get_xml_title(title.text).lower()
+                for label_section in SplHistoricalLabels.LABEL_SECTIONS
+            )
+
+            # test if any string matches is in LABEL_SECTIONS_EXACT
+            title_string_in_LABEL_SECTION_EXACT = any(
+                label_section.lower() == get_xml_title(title.text).lower()
+                for label_section in SplHistoricalLabels.LABEL_SECTIONS_EXACT
+            )
+
+            # test if any substring is in LABEL_SECTION_BLACKLIST
+            title_substring_in_LABEL_SECTION_BLACKLIST = any(
```

Comment:
Why is `LABEL_SECTIONS_BLACKLIST` required? Is it not enough to check the requires labels list?

Reply:
Maybe there might be a better way to do this without blacklists, but `Indications` from LABEL_SECTIONS would naturally cause `Contraindications`, which should be excluded from the sections of interest as identified by Dr. Marman, to be included with the sections collected.  I want to keep the number with the section title.  So if the section title is '2 Indications', I want to keep the entirety of the section title.  The number is useful for diffing as it can be used to track where new sections are added or removed relative to other sections.  The best way that I came up to do this is to test if the substring 'indications" is in the title.  However 'indications' is also in 'contraindications'.   To exclude contraindications,  a blacklist is used.

Reply:
@TerryChau I've refactored the code a little bit. And I've verified that the result is the same with your version of the code and my latest commit, for the example set ids you had quoted.
```
    "762b51be-1893-4cd1-9511-e645fc420d3a",
    "1b5e2860-6855-4a65-8bbc-e064172a1adf",
    "b5cee013-000f-4e35-a284-1f58add31b4d"
```

The only difference is with Set ID `762b51be-1893-4cd1-9511-e645fc420d3a`. The sub-section `6.1 Clinical Trials Experience` was being matched because of the string contains comparison. I've changed this up to use an 'exact match', after cleaning the text:

```
# For matching:
            # * Convert title text to upper case
            # * Replace & with AND, tab with space
            # * Clean up Unicode characters, digits, punctuation
            title_match_text = cleantext.clean(
                title.text.upper().replace("&", "AND").replace("\t", " "),
                lower=False,
                no_line_breaks=True,
                no_punct=True,
                no_numbers=True,
                no_digits=True,
                replace_with_number="",
                replace_with_digit="",
            )
```